src/internal.rs

```diff
@@ -1062,18 +1092,31 @@ pub struct Rav1dTileState {
     // each entry is one tile-sbrow; middle index is refidx
     pub lowest_pixel: usize,
 
-    pub dqmem: [[[u16; 2]; 3]; RAV1D_MAX_SEGMENTS as usize], /* [RAV1D_MAX_SEGMENTS][3 plane][2 dc/ac] */
-    pub dq: TileStateRef,
-    pub last_qidx: u8,
-    pub last_delta_lf: [i8; 4],
-    pub lflvlmem: [[[[u8; 2]; 8]; 4]; 8], /* [8 seg_id][4 dir][8 ref][2 is_gmv] */
-    pub lflvl: TileStateRef,
+    pub dqmem: RwLock<[[[u16; 2]; 3]; RAV1D_MAX_SEGMENTS as usize]>, /* [RAV1D_MAX_SEGMENTS][3 plane][2 dc/ac] */
```

Comment:
This is passed to `init_quant_tables`, where the elements being `AtomicU16`s instead of `u16`s doesn't make codegen worse (e.x. it can't combine 2`u16` stores into a `u32` store), so if we do it this way, there'll be no perf/memory overhead, like with the `RwLock`.

Note that the similar `lflvlmem` is different.  `AtomicU8`s there make codegen worse in `rav1d_calc_lf_values`.

Reply:
Making this atomic will require also making `Rav1dFrameData.dq` atomic, since we have code that conditionally operates on either `f.dq` or `ts.dqmem`. That shouldn't be an issue in this case though, because `f.dq` is used in pretty much exactly the same way as `dqmem`. I've updated both to make them atomic.